pkg/operator/ozone/controller.go

```diff
@@ -0,0 +1,280 @@
+
+import (
+	"bytes"
+	"github.com/coreos/pkg/capnslog"
+	"github.com/elek/flekszible/api"
+	"github.com/elek/flekszible/api/data"
+	"github.com/elek/flekszible/api/processor"
+	opkit "github.com/rook/operator-kit"
+	ozonev1alpha1 "github.com/rook/rook/pkg/apis/ozone.rook.io/v1alpha1"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/operator/k8sutil"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/tools/cache"
+	"os"
+	"reflect"
+)
+
+const (
+	customResourceName       = "objectstore"
+	customResourceNamePlural = "objectstores"
+)
+
+var logger = capnslog.NewPackageLogger("github.com/rook/rook", "ozone-op-object")
+
+// ObjectStoreResource represents the object store custom resource
+var ObjectStoreResource = opkit.CustomResource{
+	Name:    customResourceName,
+	Plural:  customResourceNamePlural,
+	Group:   ozonev1alpha1.CustomResourceGroup,
+	Version: ozonev1alpha1.Version,
+	Scope:   apiextensionsv1beta1.NamespaceScoped,
+	Kind:    reflect.TypeOf(ozonev1alpha1.ObjectStore{}).Name(),
+}
+
+// Controller represents a controller object for object store custom resources
+type Controller struct {
+	context     *clusterd.Context
+	templateDir string
+}
+
+// NewController create controller for watching object store custom resources created
+func NewController(context *clusterd.Context) *Controller {
+	templateDir := os.Getenv("ROOK_OZONE_TEMPLATE_DIR")
+	if templateDir == "" {
+		templateDir = "/var/lib/ozone-templates"
+	}
+
+	return &Controller{
+		context:     context,
+		templateDir: templateDir,
+	}
+}
+
+// StartWatch watches for instances of ObjectStore custom resources and acts on them
+func (c *Controller) StartWatch(namespace string, stopCh chan struct{}) error {
+	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	}
+
+	logger.Infof("start watching object store resources in namespace %s", namespace)
+	watcher := opkit.NewWatcher(ObjectStoreResource, namespace, resourceHandlerFuncs, c.context.RookClientset.OzoneV1alpha1().RESTClient())
+	go watcher.Watch(&ozonev1alpha1.ObjectStore{}, stopCh)
+
+	return nil
+}
+
+//The main resource file generation logic from the CRD
+func (c *Controller) generateResources(objectStore *ozonev1alpha1.ObjectStore) error {
+	context, err := flekszible.Initialize(c.templateDir)
+	if err != nil {
+		return err
+	}
+
+	processors := make([]processor.Processor, 0)
+
+	processors = append(processors, &processor.Namespace{
+		Namespace:          objectStore.Namespace,
+		Force:              true,
+		ClusterRoleSupport: true,
+	})
+
+	processors = append(processors, &processor.Image{
+		Image: objectStore.Spec.OzoneVersion.Image,
```

Comment:
In general, you would add the CRD settings here to be replaced in the templates?

Reply:
Do you anticipate custom management from the ozone operator in the future? Right now it seems like a glorified helm chart. :)

Reply:
Very good question: why do we need operator if the application itself is cloud-native. I tried to answer the question in the design doc, but it's a theoretical question I don't know if there is good, generic answer. I think one added value here is to adapt the hard disk detection and make it easy to configure the devices which would require manual settings with helm/kustomize anyway. Generally everything which depends from the runtime environment can not be handled by helm chart.

From technical point of view:

The current templates are copied from the ozone release and can be read from the apache/ozone docker image. The main idea is to merge the two feature set: the main resource set an be generated with the Ozone way and all the operator customization can be added with the Rook way.

The first rook style modification should be the usage of hostPath, IMHO.